### ISSUE TITLE
hanging L2 to L1 withdrawals when using `withdraw` or `withdrawTo`

### ISSUE THREAD

**Issue by platocrat** [ethereum-optimism/optimism#849]
**Describe the bug**
This issue arose from Ale's integration tests.

When running all integration tests for `synthetix`'s `develop` branch, the tests will always hang at the first L2 to L1 withdrawal step. The individual test for this step is named "when waiting for the tx to complete on L1".

**To Reproduce**
1. Git clone `synthetix`: `git clone https://github.com/Synthetixio/synthetix.git`
2. Install deps: `npm run clean-install`
3. Deploy contracts: `node publish deploy-ovm-pair`
3. Run integration tests: `npm run test:prod:ovm`
4. See the tests first hang on the withdrawals test, "when waiting for the tx to complete on L1":
<img width="751" alt="Screen Shot 2021-05-12 at 9 30 22 AM" src="https://user-images.githubusercontent.com/37757724/117992594-adad7800-b304-11eb-8fe5-180d0a1931e4.png">

5. Restart the relayer and batch submitter using the Docker Desktop dashboard/UI:
<img width="1243" alt="Screen Shot 2021-05-12 at 9 31 09 AM" src="https://user-images.githubusercontent.com/37757724/117992770-ca49b000-b304-11eb-8d74-d4d4ae9702c0.png">

6. See tests continue to pass

**Expected behavior**

**Screenshots**
The screenshots are below are of the container logs for `ops_relayer_1` and `op_batch_submitter_1` as the tests begin to hang:
1. `ops_relayer_1` logs:
<img width="1247" alt="Screen Shot 2021-05-12 at 9 38 15 AM" src="https://user-images.githubusercontent.com/37757724/117993949-c9654e00-b305-11eb-9484-979cad1293de.png">
The pattern of repeated logs continue to repeat until the relay is restarted.

2. `ops_batch_submitter_1` logs:
<img width="1247" alt="Screen Shot 2021-05-12 at 9 44 16 AM" src="https://user-images.githubusercontent.com/37757724/117996332-af2c6f80-b307-11eb-8124-236ed4b07e46.png">
No new batches are submitted before tests begin to hang. So, these logs are static once the tests are hanging.

Successful logs of the relayer after the container is restarted:
1. `ops_relayer_1` logs:
<img width="1247" alt="Screen Shot 2021-05-12 at 9 49 50 AM" src="https://user-images.githubusercontent.com/37757724/117996221-958b2800-b307-11eb-9055-7ee92a201b4b.png">

**System Specs:**
 - OS: macOS
 - Package Version: :

**Additional context**
If you don't restart the batch submitter nor relayer containers, the "before all" hook for [WITHDRAW]'s "when waiting for the tx to complete on L1" will timeout.
Tests will continue to pass as expected afterward until reaching the same test for [WITHDRAW TO], where these test will also timeout in the "before all" hook.

<img width="902" alt="Screen Shot 2021-05-12 at 10 07 58 AM" src="https://user-images.githubusercontent.com/37757724/117998901-f0258380-b309-11eb-84e5-846007f39d6d.png">

-----
**Note**
It's difficult to get both the batch submitter logs near the same time as the relayer logs because the `ops_batch_submitter_1` logs only seem to update when backing out of and returning to the `LOGS` tab in the UI.

**Comment by tynes** [comment]
The logline in `ops_relayer_1` that says `Did not find L2->L1 messages` is suspect. You can get the combined logs with:

```
docker-compose logs op_relayer_1 ops_batch_submitter_1
```

or

```
docker-compose logs | egrep 'batch_submitter|relayer'
```

**Comment by smartcontracts** [comment]
Taking a look now

**Comment by smartcontracts** [comment]
Unable to reproduce against `develop`. Here's my only test failure:

```
[0]       [WITHDRAW] when withdrawing SNX from L2 to L1
[0]         when waiting for the tx to complete on L2
[0]           when a user has the expected amount of SNX in L2
[0]             ✓ shows the user has SNX
[0]             when the SNX rate has been updated
[0]               when initiation is suspended on L2
[0]                 ✓ reverts when the user attempts to initiate a withdrawal
[0]               before a user initiates a withdrawal on L2
[0]                 ✓ inititation is active on both layers
[0]                 when a user initiates a withdrawal on L2
[0]                   ✓ reduces the users balance
[0]                   ✓ inititation is deactivated on L1
[0]                   when waiting for the tx to complete on L1
[0]                     1) emitted a WithdrawalFinalized event
[0]                     ✓ shows that the users L1 balance increased
[0]                     ✓ shows that the escrow balance decreased
``

**Comment by smartcontracts** [comment]
https://asciinema.org/a/5kJA5oM4jjc1IetH9rW34p3pT

**Comment by K-Ho** [comment]
cc @ajsantander

**Comment by platocrat** [comment]
Here are the full logs of the batch submitter and relayer while I ran these tests again (still hanging) and had to restart both services:

[bs-and-relayer-logs-of-synthetix-tests.zip](https://github.com/ethereum-optimism/optimism/files/6496579/bs-and-relayer-logs-of-synthetix-tests.zip)

I also added a few comments, that you can find by searching for "---------" (number of `-`s matters), that I think may help

**Comment by platocrat** [comment]
Ale from Synthetix ran a few experiments that failed:
> 1. Spun up `ops`, waited 10 minutes, deployed the Synthetix instances, waited 10 minutes, and only then started the tests. Tests timed out in withdrawals.
> 2. Let the test run for an hour during a "hang". No changes.
> 3. Tried sending transactions on both chains during a hang every few seconds, to see if it revived the watcher. Nothing...
>
> Before it was just the withdrawal finalization watchers, now its any finalization watcher

cc: @smartcontracts @tynes

**Comment by smartcontracts** [comment]
@platocrat I'm away from my desktop until late tomorrow and I have a hard time spinning up the whole system on my laptop. I'll get back to this on Weds.

**Comment by smartcontracts** [comment]
Hmmm those logs are quite interesting. Especially:

```
----------------- RELAYER and BATCH SUBMITTER ARE RESTARTED -----------------

relayer_1            | {"level":30,"time":1621277317406,"msg":"Service is starting..."}
relayer_1            | {"level":30,"time":1621277317408,"msg":"Service is initializing..."}
```

cc @tynes why would the relayer be restarting? Is it getting OOM killed?

**Comment by platocrat** [comment]
> Hmmm those logs are quite interesting. Especially:
> 
> ```
> ----------------- RELAYER and BATCH SUBMITTER ARE RESTARTED -----------------
> 
> relayer_1            | {"level":30,"time":1621277317406,"msg":"Service is starting..."}
> relayer_1            | {"level":30,"time":1621277317408,"msg":"Service is initializing..."}
> ```
> 
> cc @tynes why would the relayer be restarting? Is it getting OOM killed?

I initiated the restart lmao

**Comment by platocrat** [comment]
@K-Ho is I know Ale is blocked on this, is this also blocking Synthetix collectively?
I'm wondering how to move forward with this error

**Comment by K-Ho** [comment]
This is also starting to block our CI - see timeout withdrawal: https://github.com/ethereum-optimism/optimism/pull/917/checks?check_run_id=2623010867

**Comment by smartcontracts** [comment]
I believe this is now fixed as of #926 :-) going to close unless anyone else starts to see this issue
